File: library/src/androidTest/java/com/google/maps/android/data/geojson/GeoJsonParserTest.java
/*
 * Copyright 2020 Google Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.maps.android.data.geojson;

import com.google.android.gms.maps.model.LatLng;
import com.google.maps.android.data.Geometry;

import org.json.JSONObject;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.google.maps.android.data.geojson.GeoJsonTestUtil.emptyFile;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureCollectionNoCoordinatesInGeometryInFeature;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureCollectionNoFeaturesArray;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureCollectionNoGeometryTypeInFeature;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureCollectionNoTypeInGeometryInFeature;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureGeometryCollectionInvalidGeometry;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureGeometryCollectionNoGeometries;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureMissingType;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureNoCoordinatesInGeometry;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureNoGeometry;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidFeatureNoProperties;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidGeometryCollectionInvalidGeometries;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidGeometryCollectionInvalidGeometry;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidGeometryInvalidCoordinatesPair;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidGeometryInvalidCoordinatesString;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidGeometryNoCoordinates;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.invalidGeometryNoType;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.validGeometryCollection;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.validLineString;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.validMultiLineString;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.validMultiPoint;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.validMultiPolygon;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.validPoint;
import static com.google.maps.android.data.geojson.GeoJsonTestUtil.validPolygon;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

public class GeoJsonParserTest {
    @Test
    public void testParseGeoJson() throws Exception {
        JSONObject validGeoJsonObject =
                new JSONObject(
                        "{ \"type\": \"MultiLineString\",\n"
                                + "    \"coordinates\": [\n"
                                + "        [ [100.0, 0.0], [101.0, 1.0] ],\n"
                                + "        [ [102.0, 2.0], [103.0, 3.0] ]\n"
                                + "      ]\n"
                                + "    }");

        GeoJsonParser parser = new GeoJsonParser(validGeoJsonObject);
        GeoJsonLineString ls1 =
                new GeoJsonLineString(
                        new ArrayList<>(Arrays.asList(new LatLng(0, 100), new LatLng(1, 101))));
        GeoJsonLineString ls2 =
                new GeoJsonLineString(
                        new ArrayList<>(Arrays.asList(new LatLng(2, 102), new LatLng(3, 103))));
        GeoJsonMultiLineString geoJsonMultiLineString =
                new GeoJsonMultiLineString(new ArrayList<>(Arrays.asList(ls1, ls2)));
        GeoJsonFeature geoJsonFeature =
                new GeoJsonFeature(geoJsonMultiLineString, null, null, null);
        List<GeoJsonFeature> geoJsonFeatures = new ArrayList<>(Arrays.asList(geoJsonFeature));
        assertEquals(geoJsonFeatures.get(0).getId(), parser.getFeatures().get(0).getId());
    }

    @Test
    public void testParseGeometryCollection() throws Exception {
        GeoJsonParser parser = new GeoJsonParser(validGeometryCollection());
        assertEquals(1, parser.getFeatures().size());
        for (GeoJsonFeature feature : parser.getFeatures()) {
            assertEquals("GeometryCollection", feature.getGeometry().getGeometryType());
            int size = 0;
            for (String property : feature.getPropertyKeys()) {
                size++;
            }
            assertEquals(2, size);
            assertEquals("Popsicles", feature.getId());
            GeoJsonGeometryCollection geometry =
                    ((GeoJsonGeometryCollection) feature.getGeometry());
            assertEquals(1, geometry.getGeometries().size());
            for (Geometry geoJsonGeometry : geometry.getGeometries()) {
                assertEquals("GeometryCollection", geoJsonGeometry.getGeometryType());
            }
        }
    }

    @Test
    public void testParsePoint() throws Exception {
        GeoJsonParser parser = new GeoJsonParser(validPoint());
        assertEquals(1, parser.getFeatures().size());
        assertNull(parser.getFeatures().get(0).getBoundingBox());
        assertNull(parser.getFeatures().get(0).getId());
        int size = 0;
        for (String property : parser.getFeatures().get(0).getPropertyKeys()) {
            size++;
        }
        assertEquals(0, size);
        assertTrue(parser.getFeatures().get(0).getGeometry() instanceof GeoJsonPoint);
        GeoJsonPoint point = (GeoJsonPoint) parser.getFeatures().get(0).getGeometry();
        assertEquals(new LatLng(0.0, 100.0), point.getCoordinates());
    }

    @Test
    public void testParseLineString() throws Exception {
        GeoJsonParser parser = new GeoJsonParser(validLineString());
        assertEquals(1, parser.getFeatures().size());
        assertNull(parser.getFeatures().get(0).getBoundingBox());
        assertNull(parser.getFeatures().get(0).getId());
        int size = 0;
        for (String property : parser.getFeatures().get(0).getPropertyKeys()) {
            size++;
        }
        assertEquals(0, size);
        assertTrue(parser.getFeatures().get(0).getGeometry() instanceof GeoJsonLineString);
        GeoJsonLineString lineString =
                (GeoJsonLineString) parser.getFeatures().get(0).getGeometry();
        assertEquals(2, lineString.getCoordinates().size());
        List<LatLng> ls = new ArrayList<>(Arrays.asList(new LatLng(0, 100), new LatLng(1, 101)));
        assertEquals(ls, lineString.getCoordinates());
    }

    @Test
    public void testParsePolygon() throws Exception {
        GeoJsonParser parser = new GeoJsonParser(validPolygon());
        assertEquals(1, parser.getFeatures().size());
        assertNull(parser.getFeatures().get(0).getBoundingBox());
        assertNull(parser.getFeatures().get(0).getId());
        int size = 0;
        for (String property : parser.getFeatures().get(0).getPropertyKeys()) {
            size++;
        }
        assertEquals(0, size);
        assertTrue(parser.getFeatures().get(0).getGeometry() instanceof GeoJsonPolygon);
        GeoJsonPolygon polygon = (GeoJsonPolygon) parser.getFeatures().get(0).getGeometry();
        assertEquals(2, polygon.getCoordinates().size());
        assertEquals(5, polygon.getCoordinates().get(0).size());
        assertEquals(5, polygon.getCoordinates().get(1).size());
        List<ArrayList<LatLng>> p = new ArrayList<>();
        p.add(
                new ArrayList<>(
                        Arrays.asList(
                                new LatLng(0, 100),
                                new LatLng(0, 101),
                                new LatLng(1, 101),
                                new LatLng(1, 100),
                                new LatLng(0, 100))));
        p.add(
                new ArrayList<>(
                        Arrays.asList(
                                new LatLng(0.2, 100.2),
                                new LatLng(0.2, 100.8),
                                new LatLng(0.8, 100.8),
                                new LatLng(0.8, 100.2),
                                new LatLng(0.2, 100.2))));
        assertEquals(p, polygon.getCoordinates());
    }

    @Test
    public void testParseMultiPoint() throws Exception {
        GeoJsonParser parser = new GeoJsonParser(validMultiPoint());
        assertEquals(1, parser.getFeatures().size());
        assertNull(parser.getFeatures().get(0).getBoundingBox());
        assertNull(parser.getFeatures().get(0).getId());
        int size = 0;
        for (String property : parser.getFeatures().get(0).getPropertyKeys()) {
            size++;
        }
        assertEquals(0, size);
        assertTrue(parser.getFeatures().get(0).getGeometry() instanceof GeoJsonMultiPoint);
        GeoJsonMultiPoint multiPoint =
                (GeoJsonMultiPoint) parser.getFeatures().get(0).getGeometry();
        assertEquals(2, multiPoint.getPoints().size());
        assertEquals(new LatLng(0, 100), multiPoint.getPoints().get(0).getCoordinates());
        assertEquals(new LatLng(1, 101), multiPoint.getPoints().get(1).getCoordinates());
    }

    @Test
    public void testParseMultiLineString() throws Exception {
        GeoJsonParser parser = new GeoJsonParser(validMultiLineString());
        assertEquals(1, parser.getFeatures().size());
        assertNull(parser.getFeatures().get(0).getBoundingBox());
        assertNull(parser.getFeatures().get(0).getId());
        int size = 0;
        for (String property : parser.getFeatures().get(0).getPropertyKeys()) {
            size++;
        }
        assertEquals(0, size);
        assertTrue(parser.getFeatures().get(0).getGeometry() instanceof GeoJsonMultiLineString);
        GeoJsonMultiLineString multiLineString =
                (GeoJsonMultiLineString) parser.getFeatures().get(0).getGeometry();
        assertEquals(2, multiLineString.getLineStrings().size());
        List<LatLng> ls = new ArrayList<>(Arrays.asList(new LatLng(0, 100), new LatLng(1, 101)));
        assertEquals(ls, multiLineString.getLineStrings().get(0).getCoordinates());
        ls = new ArrayList<>(Arrays.asList(new LatLng(2, 102), new LatLng(3, 103)));
        assertEquals(ls, multiLineString.getLineStrings().get(1).getCoordinates());
    }

    @Test
    public void testParseMultiPolygon() throws Exception {
        GeoJsonParser parser = new GeoJsonParser(validMultiPolygon());
        assertEquals(1, parser.getFeatures().size());
        GeoJsonFeature feature = parser.getFeatures().get(0);
        GeoJsonMultiPolygon polygon = ((GeoJsonMultiPolygon) feature.getGeometry());
        assertEquals(2, polygon.getPolygons().size());
        assertEquals(1, polygon.getPolygons().get(0).getCoordinates().size());
        List<List<LatLng>> p1 = new ArrayList<>();
        p1.add(
                new ArrayList<>(
                        Arrays.asList(
                                new LatLng(2, 102),
                                new LatLng(2, 103),
                                new LatLng(3, 103),
                                new LatLng(3, 102),
                                new LatLng(2, 102))));
        assertEquals(p1, polygon.getPolygons().get(0).getCoordinates());
        assertEquals(2, polygon.getPolygons().get(1).getCoordinates().size());
        List<List<LatLng>> p2 = new ArrayList<>();
        p2.add(
                new ArrayList<>(
                        Arrays.asList(
                                new LatLng(0, 100),
                                new LatLng(0, 101),
                                new LatLng(1, 101),
                                new LatLng(1, 100),
                                new LatLng(0, 100))));
        p2.add(
                new ArrayList<>(
                        Arrays.asList(
                                new LatLng(0.2, 100.2),
                                new LatLng(0.2, 100.8),
                                new LatLng(0.8, 100.8),
                                new LatLng(0.8, 100.2),
                                new LatLng(0.2, 100.2))));
        assertEquals(p2, polygon.getPolygons().get(1).getCoordinates());
    }

    @Test
    public void testEmptyFile() {
        GeoJsonParser parser = new GeoJsonParser(emptyFile());
        assertNull(parser.getBoundingBox());
        assertEquals(0, parser.getFeatures().size());
    }

    @Test
    public void testInvalidFeature() throws Exception {
        GeoJsonParser parser;

        // Feature exists without geometry
        parser = new GeoJsonParser(invalidFeatureNoGeometry());
        assertNull(parser.getBoundingBox());
        assertEquals(1, parser.getFeatures().size());
        assertNull(parser.getFeatures().get(0).getGeometry());
        assertEquals("Dinagat Islands", parser.getFeatures().get(0).getProperty("name"));

        // Feature exists without properties
        parser = new GeoJsonParser(invalidFeatureNoProperties());
        assertNull(parser.getBoundingBox());
        assertEquals(1, parser.getFeatures().size());
        int size = 0;
        for (String property : parser.getFeatures().get(0).getPropertyKeys()) {
            size++;
        }
        assertEquals(0, size);

        // No features exist due to no type
        parser = new GeoJsonParser(invalidFeatureMissingType());
        assertNull(parser.getBoundingBox());
        assertEquals(0, parser.getFeatures().size());

        // 1 geometry in geometry collection, other geometry was invalid
        parser = new GeoJsonParser(invalidFeatureGeometryCollectionInvalidGeometry());
        assertNull(parser.getBoundingBox());
        assertEquals(1, parser.getFeatures().size());
        assertEquals(
                1,
                ((GeoJsonGeometryCollection) parser.getFeatures().get(0).getGeometry())
                        .getGeometries()
                        .size());

        // No geometry due to no geometries array member
        parser = new GeoJsonParser(invalidFeatureGeometryCollectionNoGeometries());
        assertNull(parser.getBoundingBox());
        assertEquals(1, parser.getFeatures().size());
        assertNull(parser.getFeatures().get(0).getGeometry());
    }

    @Test
    public void testInvalidFeatureCollection() throws Exception {
        GeoJsonParser parser;

        // 1 feature without geometry
        parser = new GeoJsonParser(invalidFeatureCollectionNoCoordinatesInGeometryInFeature());
        assertNull(parser.getBoundingBox());
        assertEquals(3, parser.getFeatures().size());
        assertNotNull(parser.getFeatures().get(0).getGeometry());
        assertNull(parser.getFeatures().get(1).getGeometry());
        assertNotNull(parser.getFeatures().get(2).getGeometry());

        // 1 feature without geometry
        parser = new GeoJsonParser(invalidFeatureCollectionNoTypeInGeometryInFeature());
        assertNull(parser.getBoundingBox());
        assertEquals(3, parser.getFeatures().size());
        assertNotNull(parser.getFeatures().get(0).getGeometry());
        assertNotNull(parser.getFeatures().get(1).getGeometry());
        assertNull(parser.getFeatures().get(2).getGeometry());

        // No features due to no features array
        parser = new GeoJsonParser(invalidFeatureCollectionNoFeaturesArray());
        assertNull(parser.getBoundingBox());
        assertEquals(0, parser.getFeatures().size());

        // 1 feature not parsed due to no type indicating it is a feature
        parser = new GeoJsonParser(invalidFeatureCollectionNoGeometryTypeInFeature());
        assertNull(parser.getBoundingBox());
        assertEquals(2, parser.getFeatures().size());
        assertTrue(
                !parser.getFeatures().get(0).getGeometry().getGeometryType().equals("Polygon")
                        && !parser.getFeatures()
                        .get(1)
                        .getGeometry()
                        .getGeometryType()
                        .equals("Polygon"));

        // Contains 1 feature element with no geometry as it was missing a coordinates member
        parser = new GeoJsonParser(invalidFeatureNoCoordinatesInGeometry());
        assertNull(parser.getBoundingBox());
        assertEquals(1, parser.getFeatures().size());
        assertNull(parser.getFeatures().get(0).getGeometry());
    }

    @Test
    public void testInvalidGeometry() throws Exception {
        GeoJsonParser parser;

        // No geometry due to no type member
        parser = new GeoJsonParser(invalidGeometryNoType());
        assertNull(parser.getBoundingBox());
        assertEquals(0, parser.getFeatures().size());

        // No geometry due to no coordinates member
        parser = new GeoJsonParser(invalidGeometryNoCoordinates());
        assertNull(parser.getBoundingBox());
        assertEquals(0, parser.getFeatures().size());

        // Geometry collection has 1 valid and 1 invalid geometry
        parser = new GeoJsonParser(invalidGeometryCollectionInvalidGeometry());
        assertNull(parser.getBoundingBox());
        assertEquals(1, parser.getFeatures().size());
        assertEquals(
                1,
                ((GeoJsonGeometryCollection) parser.getFeatures().get(0).getGeometry())
                        .getGeometries()
                        .size());

        // No geometry due to invalid geometry collection
        parser = new GeoJsonParser(invalidGeometryCollectionInvalidGeometries());
        assertNull(parser.getBoundingBox());
        assertEquals(0, parser.getFeatures().size());

        // No geometry due to only lng provided
        parser = new GeoJsonParser(invalidGeometryInvalidCoordinatesPair());
        assertEquals(0, parser.getFeatures().size());

        // No geometry due to coordinates being strings instead of doubles
        parser = new GeoJsonParser(invalidGeometryInvalidCoordinatesString());
        assertEquals(0, parser.getFeatures().size());
    }
}

File: onion/src/test/java/com/google/maps/android/data/parser/geojson/GeoJsonParserTest.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.maps.android.data.parser.geojson

import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.ByteArrayInputStream
import java.io.File
import kotlin.test.assertFailsWith

@RunWith(RobolectricTestRunner::class)
class GeoJsonParserTest {

    private val parser = GeoJsonParser()

    @Test
    fun testParsePoint() {
        val geoJson = """
            {
              "type": "FeatureCollection",
              "features": [
                {
                  "type": "Feature",
                  "geometry": {
                    "type": "Point",
                    "coordinates": [102.0, 0.5]
                  },
                  "properties": {
                    "prop0": "value0"
                  }
                }
              ]
            }
        """.trimIndent()
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val geoJsonObj = parser.parse(stream) as GeoJsonFeatureCollection

        assertThat(geoJsonObj.features.size).isEqualTo(1)
        val feature = geoJsonObj.features[0]
        assertThat(feature.properties?.size).isEqualTo(1)
        assertThat(feature.properties?.get("prop0")).isEqualTo("value0")
        val point = feature.geometry as GeoJsonPoint
        assertThat(point.coordinates).isEqualTo(Coordinates(0.5, 102.0))
    }

    @Test
    fun testParseLineString() {
        val geoJson = """
            {
              "type": "FeatureCollection",
              "features": [
                {
                  "type": "Feature",
                  "geometry": {
                    "type": "LineString",
                    "coordinates": [ [102.0, 0.0], [103.0, 1.0], [104.0, 0.0], [105.0, 1.0] ]
                  },
                  "properties": {
                    "prop0": "value0",
                    "prop1": "0.0"
                  }
                }
              ]
            }
        """.trimIndent()
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val geoJsonObj = parser.parse(stream) as GeoJsonFeatureCollection

        assertThat(geoJsonObj.features.size).isEqualTo(1)
        val feature = geoJsonObj.features[0]
        val lineString = feature.geometry as GeoJsonLineString
        assertThat(lineString.coordinates.size).isEqualTo(4)
        assertThat(lineString.coordinates[0]).isEqualTo(Coordinates(0.0, 102.0))
        assertThat(lineString.coordinates[1]).isEqualTo(Coordinates(1.0, 103.0))
        assertThat(lineString.coordinates[2]).isEqualTo(Coordinates(0.0, 104.0))
        assertThat(lineString.coordinates[3]).isEqualTo(Coordinates(1.0, 105.0))
    }

    @Test
    fun testParsePolygon() {
        val geoJson = """
            {
              "type": "FeatureCollection",
              "features": [
                {
                  "type": "Feature",
                  "geometry": {
                    "type": "Polygon",
                    "coordinates": [
                      [ [100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0] ]
                    ]
                  }
                }
              ]
            }
        """.trimIndent()
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val geoJsonObj = parser.parse(stream) as GeoJsonFeatureCollection

        assertThat(geoJsonObj.features.size).isEqualTo(1)
        val feature = geoJsonObj.features[0]
        val polygon = feature.geometry as GeoJsonPolygon
        assertThat(polygon.coordinates.size).isEqualTo(1)
        assertThat(polygon.coordinates[0].size).isEqualTo(5)
    }

    @Test
    fun testParseMultiPoint() {
        val geoJson = """
        {
          "type": "Feature",
          "geometry": {
            "type": "MultiPoint",
            "coordinates": [ [100.0, 0.0], [101.0, 1.0] ]
          }
        }
        """.trimIndent()
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val geoJsonObj = parser.parse(stream) as GeoJsonFeature

        val multiPoint = geoJsonObj.geometry as GeoJsonMultiPoint
        assertThat(multiPoint.coordinates.size).isEqualTo(2)
        assertThat(multiPoint.coordinates[0]).isEqualTo(Coordinates(0.0, 100.0))
        assertThat(multiPoint.coordinates[1]).isEqualTo(Coordinates(1.0, 101.0))
    }

    @Test
    fun testParseMultiLineString() {
        val geoJson = """
        {
          "type": "Feature",
          "geometry": {
            "type": "MultiLineString",
            "coordinates": [
              [ [100.0, 0.0], [101.0, 1.0] ],
              [ [102.0, 2.0], [103.0, 3.0] ]
            ]
          }
        }
        """.trimIndent()
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val geoJsonObj = parser.parse(stream) as GeoJsonFeature

        val multiLineString = geoJsonObj.geometry as GeoJsonMultiLineString
        assertThat(multiLineString.coordinates.size).isEqualTo(2)
        assertThat(multiLineString.coordinates[0][0]).isEqualTo(Coordinates(0.0, 100.0))
        assertThat(multiLineString.coordinates[0][1]).isEqualTo(Coordinates(1.0, 101.0))
        assertThat(multiLineString.coordinates[1][0]).isEqualTo(Coordinates(2.0, 102.0))
        assertThat(multiLineString.coordinates[1][1]).isEqualTo(Coordinates(3.0, 103.0))
    }

    @Test
    fun testParseMultiPolygon() {
        val geoJson = """
        {
          "type": "Feature",
          "geometry": {
            "type": "MultiPolygon",
            "coordinates": [
              [[[102.0, 2.0], [103.0, 2.0], [103.0, 3.0], [102.0, 3.0], [102.0, 2.0]]],
              [[[100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0]]]
            ]
          }
        }
        """.trimIndent()
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val geoJsonObj = parser.parse(stream) as GeoJsonFeature

        val multiPolygon = geoJsonObj.geometry as GeoJsonMultiPolygon
        assertThat(multiPolygon.coordinates.size).isEqualTo(2)
    }

    @Test
    fun testParseGeometryCollection() {
        val geoJson = """
        {
          "type": "Feature",
          "geometry": {
            "type": "GeometryCollection",
            "geometries": [
              {
                "type": "Point",
                "coordinates": [100.0, 0.0]
              },
              {
                "type": "LineString",
                "coordinates": [ [101.0, 0.0], [102.0, 1.0] ]
              }
            ]
          }
        }
        """.trimIndent()
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val geoJsonObj = parser.parse(stream) as GeoJsonFeature

        assertThat(geoJsonObj.geometry).isInstanceOf(GeoJsonGeometryCollection::class.java)
        val geomCollection = geoJsonObj.geometry as GeoJsonGeometryCollection
        assertThat(geomCollection.geometries.size).isEqualTo(2)
        assertThat(geomCollection.geometries[0]).isInstanceOf(GeoJsonPoint::class.java)
        assertThat(geomCollection.geometries[1]).isInstanceOf(GeoJsonLineString::class.java)
    }

    @Test
    fun testParseNullGeometry() {
        val geoJson = """
        {
          "type": "Feature",
          "geometry": null,
          "properties": {
            "name": "null island"
          }
        }
        """.trimIndent()
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val geoJsonObj = parser.parse(stream) as GeoJsonFeature

        assertThat(geoJsonObj.geometry).isNull()
        assertThat(geoJsonObj.properties?.get("name")).isEqualTo("null island")
    }

    @Test
    fun testUsaJson() {
        val stream = File("src/test/resources/usa.json").inputStream()
        val geoJsonObj = parser.parse(stream) as GeoJsonFeatureCollection
        assertThat(geoJsonObj.features.size).isEqualTo(1)
    }

    @Test
    fun testSouthLondonSquareJson() {
        val stream = File("src/test/resources/south_london_square_geojson.json").inputStream()
        val geoJsonObj = parser.parse(stream) as GeoJsonFeatureCollection
        assertThat(geoJsonObj.features.size).isEqualTo(1)
    }

    @Test
    fun testSouthLondonLineJson() {
        val stream = File("src/test/resources/south_london_line_geojson.json").inputStream()
        val geoJsonObj = parser.parse(stream) as GeoJsonFeatureCollection
        assertThat(geoJsonObj.features.size).isEqualTo(1)
    }

    @Test
    fun testEarthquakesWithUsaJson() {
        val stream = File("src/test/resources/earthquakes_with_usa.json").inputStream()
        val geoJsonObj = parser.parse(stream) as GeoJsonFeatureCollection
        assertThat(geoJsonObj.features.isNotEmpty()).isTrue()
    }

    @Test
    fun testEarthquakesJson() {
        val stream = File("src/test/resources/earthquakes.json").inputStream()
        val geoJsonObj = parser.parse(stream) as GeoJsonFeatureCollection
        assertThat(geoJsonObj.features.isNotEmpty()).isTrue()
    }

    @Test
    fun testParseInvalidJson() {
        val geoJson = """
            {
              "type": "FeatureCollection",
              "features": [
                {
                  "type": "Feature",
                  "geometry": {
                    "type": "Point",
                    "coordinates": [102.0, 0.5]
                  },
                  "properties": {
                    "prop0": "value0"
                  }
                }
              ]
            }
        """.trimIndent().substring(10) // malformed json
        val stream = ByteArrayInputStream(geoJson.toByteArray())
        val parser = GeoJsonParser()
        assertFailsWith<Exception> {
            parser.parse(stream)
        }
    }
}